atoma-p2p/src/service.rs

```diff
@@ -0,0 +1,1125 @@
+        let gossip_message: GossipMessage = serde_cbor::from_slice(message_data)?;
+        match gossip_message {
+            GossipMessage::AddressResponse(response) => {
+                debug!(
+                    target = "atoma-p2p",
+                    event = "gossipsub_message_data",
+                    "Received gossipsub message data"
+                );
+                let message_acceptance =
+                    match utils::validate_messages(&response, self.state_manager_sender.clone())
+                        .await
+                    {
+                        Ok(_) => gossipsub::MessageAcceptance::Accept,
+                        Err(e) => {
+                            error!(
+                                target = "atoma-p2p",
+                                event = "gossipsub_message_validation_error",
+                                error = %e,
+                                "Failed to validate gossipsub message"
+                            );
+                            // NOTE: We should reject the message if it fails to validate
+                            // as it means the node is not being following the current protocol
+                            gossipsub::MessageAcceptance::Reject
+                        }
+                    };
+                // Report the message validation result to the gossipsub protocol
+                self.swarm
+                    .behaviour_mut()
+                    .gossipsub
+                    .report_message_validation_result(
+                        message_id,
+                        propagation_source,
+                        message_acceptance,
+                    )?;
+                // If the current peer is a client, we need to store the public URL in the state manager
+                if self.is_client {
+                    let event = AtomaP2pEvent::NodePublicUrlRegistrationEvent {
+                        public_url: response.address,
+                        node_small_id: response.node_small_id,
+                        timestamp: response.timestamp,
+                    };
+                    self.state_manager_sender.send((event, None)).map_err(|e| {
+                        error!(
+                            target = "atoma-p2p",
+                            event = "gossipsub_message_state_manager_error",
+                            error = %e,
+                            "Failed to send event to state manager"
+                        );
+                        AtomaP2pNodeError::StateManagerError(e)
+                    })?;
+                }
+            }
+            GossipMessage::AddressRequest => {
+                debug!(
+                    target = "atoma-p2p",
+                    event = "gossipsub_message_data",
+                    "Received gossipsub message data"
+                );
+                // Address request messages are always accepted and propagated on the p2p network
+                self.swarm
```

Comment:
@sdbondi also added a `report_message_validation_result` to always accept requests for node addresses, as if I understand correctly, otherwise this message would not be propagated to other nodes.

Reply:
That is correct - however since gossipsub gossips the message to all subscribers, you probably don't want to use it for this part of the protocol. i.e. a single request will arrive at all subscribers (i.e all nodes) and will result in all nodes sending a response to all nodes. I assume that all nodes also send a request. $O(n^2)$ message complexity.

https://docs.rs/libp2p-request-response/latest/libp2p_request_response/ might be what you're looking for